### ISSUE TITLE
Support moving tasks to different projects / sections

### ISSUE THREAD

**Issue by oscarclivio** [Doist/todoist-api-python#8]
Unless I am misunderstanding the basic principle of a REST API, I do not see any function to move a task to another project or section either in the documentation or the code.

Did I miss this method somewhere? Or can this be implemented?

**Comment by jankratochvilcz** [comment]
@oscarclivio thanks for bringing this up. Sorry for the delayed response, we're full of vacations in December

@PotHix I've tried and moving through setting the `project_id` actually works, but it's not in the REST API documentation. I know we have [move](https://developer.todoist.com/sync/v8/#move-an-item) command on the Sync API, but I'm unsure if there's any extra feature. Can you please let me know if we can support changing the `project_id` through the REST API or introduce the move command?

**Comment by PotHix** [comment]
> I've tried and moving through setting the project_id actually works

Are you sure? Via task update? That shouldn't work in the current version.

> Can you please let me know if we can support changing the project_id through the REST API or introduce the move command?

Yes, that is in the roadmap, preferably without needing a specific endpoint for the move (but needs further exploration). It will certainly get in at some point, we just don't have an ETA yet.

**Comment by jankratochvilcz** [comment]
Update: an internal ticket has been created for this, but we can't share a deadline for this. We'll post updates here when we have them.

**Comment by 5t4cktr4c3** [comment]
Any updates on this? Since the `SYNC` API v8 will be removed at the end of the month, I won't be able to use the deprecated `SYNC` python package.

**Comment by oscarclivio** [comment]
Sorry for my absence of response, it looks like I forgot I had written this issue.

It looks like the sync API is deactivated now. So any updates on this? It is a key point of my personal task management.

**Comment by 5t4cktr4c3** [comment]
My current workaround (not beautiful, but after a lot of frustrating testing I came to this) for moving a task to a new parent task:
```python
...
req = requests.post('https://api.todoist.com/sync/v9/sync', headers={'Authorization': f"Bearer {TODOIST_APITOKEN}", 'Content-Type': 'application/x-www-form-urlencoded'}, data={'commands': json.dumps([{'type': 'item_move', 'uuid': str(uuid.uuid4()), 'args': {'id': str(previousTask.id), 'parent_id': str(tasks[0].id)}}])})```
if req.status_code != 200:
	print(requests.Response.text)
...

**Comment by PotHix** [comment]
> the SYNC API v8 will be removed at the end of the month, I won't be able to use the deprecated SYNC python package.

Yes, the Python package is deprecated for a while and will not be updated for v9.

> It looks like the sync API is deactivated now. 

The Sync API is not deactivated yet. We're going to deprecate the version 8 of the Sync API, but the version 9 is already released and [has the same feature](https://developer.todoist.com/sync/v9/#move-an-item).

We know it's a bummer that we still don't have this implemented in the REST API yet, and I hope to be able to prioritize it soon.

**Comment by vp-priv** [comment]
Hm, same problem here - v8 depreciated and no python library for v9 ...
... having no possibility to move task - this "basic" method is requesting to be in REST api as I can remember

**Comment by vp-priv** [comment]
> My current workaround (not beautiful, but after a lot of frustrating testing I came to this) for moving a task to a new parent task:
> 
> ```python
> ...
> req = requests.post('https://api.todoist.com/sync/v9/sync', headers={'Authorization': f"Bearer {TODOIST_APITOKEN}", 'Content-Type': 'application/x-www-form-urlencoded'}, data={'commands': json.dumps([{'type': 'item_move', 'uuid': str(uuid.uuid4()), 'args': {'id': str(previousTask.id), 'parent_id': str(tasks[0].id)}}])})```
> if req.status_code != 200:
> 	print(requests.Response.text)
> ...
> ```

Hi, I am not pro-programmer - could you help to use your code to move task to another project? Thanks.

**Comment by 5t4cktr4c3** [comment]
```python
import requests
import uuid
import json

TODOIST_APITOKEN = None # put your token here
task = None  # get your task the way you do
project = None  # get your project the way you do

req = requests.post('https://api.todoist.com/sync/v9/sync', headers={'Authorization': f"Bearer {TODOIST_APITOKEN}", 'Content-Type': 'application/x-www-form-urlencoded'}, data={'commands': json.dumps([{'type': 'item_move', 'uuid': str(uuid.uuid4()), 'args': {'id': str(task.id), 'project_id': str(project.id)}}])})```
if req.status_code != 200:
	print(requests.Response.text)
```

That should do the trick. Note that this is untested.

**Comment by ramonvg** [comment]
A simpler approach:

```python
def move_task(task_id: str, project_id: str) -> bool:
    body = {
        "commands": [
            {
                "type": "item_move",
                "args": {"id": task_id, "project_id": project_id},
                "uuid": uuid4().hex,
            },
        ],
    }
    response = requests.post(
        "https://api.todoist.com/sync/v9/sync",
        headers={"Authorization": f"Bearer {TOKEN}"},
        json=body,
    )
    return response.ok
```

**Comment by vp-priv** [comment]
@ramonvg 
Thank you very much, I was always lost in requests and uuid, etc - it WORKS.
Vladimir

**Comment by Pa7rickStar** [comment]
Not the prettiest solution but with the tools available in Todoist REST API (not tested):

```
from todoist_api_python.api import TodoistAPI
api = TodoistAPI("0123456789abcdef0123456789")

...

def move_task(task, project_id:str=None, section_id:str=None, \
    parent_id:str=None, order:int=None):
    """'Moves' a task to a different project/ section or parent by creating a \
        identical new one and deleting the old one. 

    Args:
        task (TodoistAPI task object): task-object of the task to move.
        project_id (str, optional): Where the task should be moved to. \
            Defaults to None.
        section_id (str, optional): Where the task should be moved to. \
            Defaults to None.
        parent_id (str, optional): Where the task should be moved to. \
            Defaults to None.
        order (int, optional): Where the task should be moved to. \
            Defaults to None.

    Returns:
        task (TodoistAPI task object): If successful. Else returns None.
    """
    try:
        comments = api.get_comments(task_id=task.id)
    except Exception as error:
        print(error)
        return False
    else:
        try:
            new_task = api.add_task(
                content=task.content,
                description=task.description,
                labels=task.labels,
                priority=task.priority,
                due=task.due,
                assignee_id=task.assignee_id,
                project_id=project_id,
                section_id=section_id,
                parent_id=parent_id,
                order=order
            )
        except Exception as error:
                print(error)
                return False
        else:
            for comment in comments:
                try:
                    created_comment = api.add_comment(
                        content=comment.content,
                        task_id=task.id,
                        attachment=comment.attachment
                    )
                except Exception as error:
                    print(error)
                    return False
                else:
                    try:
                        api.delete_task(task_id=task.id)
                    except Exception as error:
                        print(error)
                        return False
                    else:
                        return new_task
```

**Comment by vp-priv** [comment]
@Pa7rickStar 
Hi, thanks - similar to one of my unsuccessful trials one year ago.
I just need to move task - any other information I have from Rest API - I cannot use create and delete task because there could be pictures, etc in the comments, and there is no functional possibility to copy them to the new task (did not work one year ago).
Vladimir

**Comment by nstielau** [comment]
@jankratochvilcz any update on REST API to move a task between projects?  If not, consider this me registering my interest ;)

**Comment by kquinsland** [comment]
Well there's 30 min i'll never get back... turns out that it's not supported, nor is there an error raised/returned to indicate as much :(.

Thanks to @ramonvg for the [workaround](https://github.com/Doist/todoist-api-python/issues/8#issuecomment-1344860782) that ... should not have been needed.

**Comment by miba87** [comment]
I also need this feature to move a task in another section. Why is this not supported?

**Comment by dylan-k** [comment]
Moving a batch of tasks from inbox to a different project was literally the first thing I tried to do with the API. Is it really not possible? That seems really surprising to me.

**Comment by nstielau** [comment]
Possible, just have to use the RPC sync API, not REST api. See
https://github.com/nstielau/todo5/blob/0dbbec85862cb9b7bd31d84bb628ca82f45f1413/src/components/InboxReviewComponent.vue#L76

**Comment by kquinsland** [comment]
> Moving a batch of tasks from inbox to a different project was literally the first thing I tried to do with the API. Is it really not possible? That seems really surprising to me.

For anything involving batches, the sync API is a more efficient/better choice anyways.

I wish the Doist team(s) would update their dev docs to make it clear that there are two APIs and that they are not 1:1 feature wise. I do not know if they changed course on deprecating the sync API or not but if that's still the plan, reflecting that in the dev docs would also be a welcome improvement.

**Comment by goncalossilva** [comment]
Having a dedicated API endpoint and a corresponding SDK method makes perfect sense. This is now on our internal backlog. I can't quite tell when we'll get around to it, but I hope it will be soon.

**Comment by goncalossilva** [comment]
There is now a dedicated API endpoint for this: https://todoist.com/api/v1/docs#tag/Tasks/operation/move_task_api_v1_tasks__task_id__move_post

Contributions welcome. 😄

**Comment by radiant-tangent** [comment]
I was going to pick up this ticket to get the new endpoint into the sdk but every time I try the `move` endpoint I get `400 Bad Request` with `Invalid argument value` in the body. I have no problem calling `get_task`, `close_task`, or `reopen_task`. Has anyone gotten this endpoint to work?
